File: piper_learnning/scripts/moveit_joint.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
import moveit_commander
import sys
import time

def main():
    # 初始化 moveit_commander 和 ROS 节点
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node("moveit_joint", anonymous=True)

    # 启动异步线程
    spinner = rospy.Rate(10)

    # 创建 MoveGroup 接口对象
    move_group = moveit_commander.MoveGroupCommander("arm")

    # 获取末端执行器的 link 名
    end_effector_link = move_group.get_end_effector_link()

    # 设置容差和规划参数
    move_group.allow_replanning(True)
    move_group.set_goal_position_tolerance(0.001)
    move_group.set_goal_orientation_tolerance(0.01)
    move_group.set_max_acceleration_scaling_factor(0.2)
    move_group.set_max_velocity_scaling_factor(0.2)

    # 控制机械臂回到初始化位置 "home"
    move_group.set_named_target("home")
    move_group.go(wait=True)
    rospy.sleep(1)

    # 获取当前的关节角度（用于补全未设置的关节）
    joint_values = move_group.get_current_joint_values()

    # 设置第 6 个关节的角度
    joint_values[5] = 0.5897185686390605

    # 设置目标关节角
    move_group.set_joint_value_target(joint_values)

    rospy.loginfo("Moving to joint target...")
    move_group.go(wait=True)

    # 关闭 MoveIt
    moveit_commander.roscpp_shutdown()

if __name__ == '__main__':
    main()

File: piper_learnning/src/moveit_joint.cpp
// 导入ROS头文件
#include <ros/ros.h>
// 导入move_group_interface头文件
#include <moveit/move_group_interface/move_group_interface.h>

// 主函数
int main(int argc, char** argv)
{
    // 初始化ROS节点
    ros::init(argc, argv, "moveit_joint");

    // 异步消息处理器，设置线程数为1
    ros::AsyncSpinner spinner(1);
    
    // 启动异步消息处理器
    spinner.start();

    // move_group接口，选择机械臂为"lite6"
	moveit::planning_interface::MoveGroupInterface move_group("arm");
    //获取终端link的名称
    std::string end_effector_link = move_group.getEndEffectorLink();
   //设置目标位置所使用的参考坐标系
    geometry_msgs::Pose start_pose = move_group.getCurrentPose(end_effector_link).pose;
    //当运动规划失败后，允许重新规划
    move_group.allowReplanning(true);

    //设置位置(单位：米)和姿态（单位：弧度）的允许误差
    move_group.setGoalPositionTolerance(0.001);
    move_group.setGoalOrientationTolerance(0.01);

    //设置允许的最大速度和加速度
    move_group.setMaxAccelerationScalingFactor(0.2);
    move_group.setMaxVelocityScalingFactor(0.2);

    // 控制机械臂先回到初始化位置
    move_group.setNamedTarget("home");
    move_group.move();
    sleep(1);

    // 声明一个double 型元素的vector容器 
    std::vector<double> joint_values(6);

    // 把关节运动位置填入vector容器 
    // joint_values[0]=-0.39441719709048184;
    // joint_values[1]= 0.4228564717254736;
    // joint_values[2]= 0.5897185686390605;
    // joint_values[3]= 0.5897185686390605;
    // joint_values[4]= 0.5897185686390605;
    joint_values[5]= 0.5897185686390605;

    // 设置运动目标
    move_group.setJointValueTarget(joint_values); 

    ROS_INFO("move");
    // 执行运动
    move_group.move(); 

    // 停止异步消息处理器
    spinner.stop();

    // 关闭ROS节点
    ros::shutdown();
    
    // 返回0，程序结束
    return 0;
}
